File: src/main/java/io/vproxy/vfx/theme/Theme.java
package io.vproxy.vfx.theme;

import io.vproxy.vfx.manager.font.FontProvider;
import io.vproxy.vfx.theme.impl.DarkTheme;
import javafx.scene.image.Image;
import javafx.scene.paint.*;

public abstract class Theme {
    private static Theme currentTheme;

    public static Theme current() {
        if (currentTheme != null) {
            return currentTheme;
        }
        synchronized (Theme.class) {
            if (currentTheme != null) {
                return currentTheme;
            }
            currentTheme = new DarkTheme();
        }
        return currentTheme;
    }

    public static void setTheme(Theme theme) {
        synchronized (Theme.class) {
            if (currentTheme != null)
                throw new IllegalStateException("currentTheme is already set");
            currentTheme = theme;
        }
    }

    public abstract FontProvider fontProvider();

    public abstract Color normalTextColor();

    public abstract Color borderColor();

    public abstract Color borderColorLight();

    public abstract Color borderColorDark();

    public Color windowBorderColor() {
        return borderColor();
    }

    public Color windowBorderColorLight() {
        return borderColorLight();
    }

    public Color windowBorderColorDark() {
        return borderColorDark();
    }

    public abstract Color sceneBackgroundColor();

    public abstract Color subSceneBackgroundColor();

    public LinearGradient makeVerticalSubSceneBackgroundGradient() {
        return new LinearGradient(0, 0, 0, 1, true,
            CycleMethod.NO_CYCLE,
            new Stop(0, Theme.current().sceneBackgroundColor()),
            new Stop(1, Theme.current().subSceneBackgroundColor()));
    }

    public abstract Image windowCloseButtonNormalImage();

    public abstract Image windowCloseButtonHoverImage();

    public abstract Image windowMaximizeButtonNormalImage();

    public abstract Image windowMaximizeButtonHoverImage();

    public abstract Image windowResetWindowSizeButtonNormalImage();

    public abstract Image windowResetWindowSizeButtonHoverImage();

    public abstract Image windowIconifyButtonNormalImage();

    public abstract Image windowIconifyButtonHoverImage();

    public Color fusionButtonNormalBackgroundColor() {
        return sceneBackgroundColor();
    }

    public abstract Color fusionButtonHoverBackgroundColor();

    public abstract Color fusionButtonDownBackgroundColor();

    public abstract Color fusionButtonAnimatingBorderLightColor();

    public Color transparentFusionButtonNormalBackgroundColor() {
        return transparentFusionPaneNormalBackgroundColor();
    }

    public Color transparentFusionButtonHoverBackgroundColor() {
        return transparentFusionPaneHoverBackgroundColor();
    }

    public abstract Color transparentFusionButtonDownBackgroundColor();

    public boolean enableFusionButtonAnimation() {
        return true;
    }

    public Color fusionPaneNormalBackgroundColor() {
        return sceneBackgroundColor();
    }

    public abstract Color fusionPaneHoverBackgroundColor();

    public abstract Color fusionPaneBorderColor();

    public Color transparentFusionPaneNormalBackgroundColor() {
        return new Color(0, 0, 0, 0);
    }

    public abstract Color transparentFusionPaneHoverBackgroundColor();

    public abstract Color scrollBarColor();

    public Color fusionButtonTextColor() {
        return normalTextColor();
    }

    public abstract Color fusionButtonDisabledTextColor();

    public abstract Color coverBackgroundColor();

    public LinearGradient makeCoverGradientBackground() {
        var color = coverBackgroundColor();
        var initial = new Color(color.getRed(), color.getGreen(), color.getBlue(), 0);
        return new LinearGradient(
            0, 0, 0, 1, true, CycleMethod.NO_CYCLE,
            new Stop(0, initial),
            new Stop(0.6, color),
            new Stop(1, color)
        );
    }

    public Color tableTextColor() {
        return normalTextColor();
    }

    public Color tableHeaderTextColor() {
        return normalTextColor();
    }

    public abstract Color tableSortLabelColor();

    public abstract Color tableCellSelectedBackgroundColor();

    public Color tableCellBackgroundColor1() {
        return sceneBackgroundColor();
    }

    public abstract Color tableCellBackgroundColor2();

    public abstract Color tableHeaderTopBackgroundColor();

    public abstract Color tableHeaderBottomBackgroundColor();

    public abstract Color progressBarProgressColor();

    public abstract Color progressBarBackgroundColor();

    public Color toggleSwitchBorderColor() {
        return borderColor();
    }

    public Color toggleSwitchUnselectedButtonColor() {
        return fusionButtonNormalBackgroundColor();
    }

    public Color toggleSwitchUnselectedButtonHoverColor() {
        return fusionButtonHoverBackgroundColor();
    }

    public Color toggleSwitchSelectedButtonColor() {
        return fusionButtonDownBackgroundColor();
    }

    public abstract Color toggleSwitchUnselectedTrayColor();

    public abstract Color toggleSwitchSelectedTrayColor();

    public Color sliderButtonNormalColor() {
        return fusionButtonNormalBackgroundColor();
    }

    public Color sliderButtonHoverColor() {
        return fusionButtonHoverBackgroundColor();
    }

    public Color sliderButtonDownColor() {
        return fusionButtonDownBackgroundColor();
    }

    public Color sliderButtonBorderColor() {
        return borderColor();
    }

    public Color rangeSliderButtonNormalColor() {
        return fusionButtonNormalBackgroundColor();
    }

    public Color rangeSliderButtonHoverColor() {
        return fusionButtonHoverBackgroundColor();
    }

    public Color rangeSliderButtonDownColor() {
        return fusionButtonDownBackgroundColor();
    }

    public Color rangeSliderButtonBorderColor() {
        return borderColor();
    }

    public Color rangeSliderBackgroundColor() {
        return progressBarBackgroundColor();
    }

    public Color rangeSliderRangeColor() {
        return progressBarProgressColor();
    }

    public Color fusionWrapperBackgroundColor() {
        return sceneBackgroundColor();
    }
}

File: src-ts/src/theme/Theme.ts
// Theme ↔ DarkTheme is a static-import cycle that crashes ESM loaders.
// The Java lazy default is removed; bootstrap always calls setTheme() first.

import type { FontProvider } from '../manager/font/FontProvider.js';
import { Color, LinearGradient } from '../javafx/color.js';
import type { FXImage } from '../javafx/ImageView.js';

export abstract class Theme {
  private static currentTheme: Theme | null = null;

  static current(): Theme {
    if (Theme.currentTheme !== null) return Theme.currentTheme;
    throw new Error('Theme.current() called before Theme.setTheme(...). Call setTheme() during bootstrap.');
  }

  static setTheme(theme: Theme): void {
    if (Theme.currentTheme !== null) {
      throw new Error('currentTheme is already set');
    }
    Theme.currentTheme = theme;
  }

  abstract fontProvider(): FontProvider;

  abstract normalTextColor(): Color;
  abstract borderColor(): Color;
  abstract borderColorLight(): Color;
  abstract borderColorDark(): Color;

  windowBorderColor(): Color { return this.borderColor(); }
  windowBorderColorLight(): Color { return this.borderColorLight(); }
  windowBorderColorDark(): Color { return this.borderColorDark(); }

  abstract sceneBackgroundColor(): Color;
  abstract subSceneBackgroundColor(): Color;

  makeVerticalSubSceneBackgroundGradient(): LinearGradient {
    return new LinearGradient(0, 0, 0, 1, true, 'NO_CYCLE', [
      { offset: 0, color: Theme.current().sceneBackgroundColor() },
      { offset: 1, color: Theme.current().subSceneBackgroundColor() },
    ]);
  }

  abstract windowCloseButtonNormalImage(): FXImage;
  abstract windowCloseButtonHoverImage(): FXImage;
  abstract windowMaximizeButtonNormalImage(): FXImage;
  abstract windowMaximizeButtonHoverImage(): FXImage;
  abstract windowResetWindowSizeButtonNormalImage(): FXImage;
  abstract windowResetWindowSizeButtonHoverImage(): FXImage;
  abstract windowIconifyButtonNormalImage(): FXImage;
  abstract windowIconifyButtonHoverImage(): FXImage;

  fusionButtonNormalBackgroundColor(): Color { return this.sceneBackgroundColor(); }
  abstract fusionButtonHoverBackgroundColor(): Color;
  abstract fusionButtonDownBackgroundColor(): Color;
  abstract fusionButtonAnimatingBorderLightColor(): Color;

  transparentFusionButtonNormalBackgroundColor(): Color {
    return this.transparentFusionPaneNormalBackgroundColor();
  }
  transparentFusionButtonHoverBackgroundColor(): Color {
    return this.transparentFusionPaneHoverBackgroundColor();
  }
  abstract transparentFusionButtonDownBackgroundColor(): Color;

  enableFusionButtonAnimation(): boolean { return true; }

  fusionPaneNormalBackgroundColor(): Color { return this.sceneBackgroundColor(); }
  abstract fusionPaneHoverBackgroundColor(): Color;
  abstract fusionPaneBorderColor(): Color;

  transparentFusionPaneNormalBackgroundColor(): Color {
    return new Color(0, 0, 0, 0);
  }
  abstract transparentFusionPaneHoverBackgroundColor(): Color;

  abstract scrollBarColor(): Color;

  fusionButtonTextColor(): Color { return this.normalTextColor(); }
  abstract fusionButtonDisabledTextColor(): Color;
  abstract coverBackgroundColor(): Color;

  makeCoverGradientBackground(): LinearGradient {
    const c = this.coverBackgroundColor();
    const initial = new Color(c.red, c.green, c.blue, 0);
    return new LinearGradient(0, 0, 0, 1, true, 'NO_CYCLE', [
      { offset: 0, color: initial },
      { offset: 0.6, color: c },
      { offset: 1, color: c },
    ]);
  }

  tableTextColor(): Color { return this.normalTextColor(); }
  tableHeaderTextColor(): Color { return this.normalTextColor(); }
  abstract tableSortLabelColor(): Color;
  abstract tableCellSelectedBackgroundColor(): Color;
  tableCellBackgroundColor1(): Color { return this.sceneBackgroundColor(); }
  abstract tableCellBackgroundColor2(): Color;
  abstract tableHeaderTopBackgroundColor(): Color;
  abstract tableHeaderBottomBackgroundColor(): Color;

  abstract progressBarProgressColor(): Color;
  abstract progressBarBackgroundColor(): Color;

  toggleSwitchBorderColor(): Color { return this.borderColor(); }
  toggleSwitchUnselectedButtonColor(): Color { return this.fusionButtonNormalBackgroundColor(); }
  toggleSwitchUnselectedButtonHoverColor(): Color { return this.fusionButtonHoverBackgroundColor(); }
  toggleSwitchSelectedButtonColor(): Color { return this.fusionButtonDownBackgroundColor(); }
  abstract toggleSwitchUnselectedTrayColor(): Color;
  abstract toggleSwitchSelectedTrayColor(): Color;

  sliderButtonNormalColor(): Color { return this.fusionButtonNormalBackgroundColor(); }
  sliderButtonHoverColor(): Color { return this.fusionButtonHoverBackgroundColor(); }
  sliderButtonDownColor(): Color { return this.fusionButtonDownBackgroundColor(); }
  sliderButtonBorderColor(): Color { return this.borderColor(); }

  rangeSliderButtonNormalColor(): Color { return this.fusionButtonNormalBackgroundColor(); }
  rangeSliderButtonHoverColor(): Color { return this.fusionButtonHoverBackgroundColor(); }
  rangeSliderButtonDownColor(): Color { return this.fusionButtonDownBackgroundColor(); }
  rangeSliderButtonBorderColor(): Color { return this.borderColor(); }
  rangeSliderBackgroundColor(): Color { return this.progressBarBackgroundColor(); }
  rangeSliderRangeColor(): Color { return this.progressBarProgressColor(); }

  fusionWrapperBackgroundColor(): Color { return this.sceneBackgroundColor(); }
}
